client/alloc_endpoint.go

```diff
@@ -100,3 +116,186 @@ func (a *Allocations) Stats(args *cstructs.AllocStatsRequest, reply *cstructs.Al
+			return
+		}
+	}
+
+	// Validate the arguments
+	if req.AllocID == "" {
+		handleStreamResultError(allocIDNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	if req.Task == "" {
+		handleStreamResultError(taskNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	if len(req.Cmd) == 0 {
+		handleStreamResultError(errors.New("command is not present"), helper.Int64ToPtr(400), encoder)
+	}
+
+	ar, err := a.c.getAllocRunner(req.AllocID)
+	if err != nil {
+		code := helper.Int64ToPtr(500)
+		if structs.IsErrUnknownAllocation(err) {
+			code = helper.Int64ToPtr(404)
+		}
+
+		handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	capabilities, err := ar.GetTaskDriverCapabilities(req.Task)
+	if err != nil {
+		code := helper.Int64ToPtr(500)
+		if structs.IsErrUnknownAllocation(err) {
+			code = helper.Int64ToPtr(404)
+		}
+
+		handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	// check node access
+	if aclObj != nil && capabilities.FSIsolation == drivers.FSIsolationNone {
+		exec := aclObj.AllowNsOp(req.QueryOptions.Namespace, acl.NamespaceCapabilityAllocNodeExec)
+		if !exec {
+			handleStreamResultError(structs.ErrPermissionDenied, nil, encoder)
+			return
+		}
+	}
+
+	allocState, err := a.c.GetAllocState(req.AllocID)
+	if err != nil {
+		code := helper.Int64ToPtr(500)
+		if structs.IsErrUnknownAllocation(err) {
+			code = helper.Int64ToPtr(404)
+		}
+
+		handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	// Check that the task is there
+	taskState := allocState.TaskStates[req.Task]
+	if taskState == nil {
+		handleStreamResultError(
+			fmt.Errorf("unknown task name %q", req.Task),
+			helper.Int64ToPtr(400),
+			encoder)
+		return
+	}
+
+	if taskState.StartedAt.IsZero() {
+		handleStreamResultError(
+			fmt.Errorf("task %q not started yet.", req.Task),
+			helper.Int64ToPtr(404),
+			encoder)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	h := ar.GetTaskExecHandler(req.Task)
+	if h == nil {
+		handleStreamResultError(
+			fmt.Errorf("task %q is not running.", req.Task),
+			helper.Int64ToPtr(404),
+			encoder)
+		return
+	}
+
+	err = h(ctx, req.Cmd, req.Tty, newExecStream(cancel, decoder, encoder))
+	if err != nil {
+		code := helper.Int64ToPtr(500)
+		handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	return
+}
+
+func newExecStream(cancelFn func(), decoder *codec.Decoder, encoder *codec.Encoder) drivers.ExecTaskStream {
+	buf := new(bytes.Buffer)
+	return &execStream{
+		cancelFn: cancelFn,
+		decoder:  decoder,
+
+		buf:        buf,
+		encoder:    encoder,
+		frameCodec: codec.NewEncoder(buf, structs.JsonHandle),
+	}
+}
+
+type execStream struct {
+	cancelFn func()
+	decoder  *codec.Decoder
+
+	encoder    *codec.Encoder
+	buf        *bytes.Buffer
+	frameCodec *codec.Encoder
+}
+
+func (s *execStream) Send(m *drivers.ExecTaskStreamingResponseMsg) error {
+	s.buf.Reset()
+	s.frameCodec.Reset(s.buf)
+
+	s.frameCodec.MustEncode(m)
```

Comment:
Shouldn't we return an error instead of panicking here?

Reply:
The only way I can reason about this failing is if the ExecTaskStreamingResponseMsg type is un-encodable by json which should never happen.  And plain `Encode` function has some overhead in setting up panic handler, but it's a micro-optimization.